pontoon/sync/tests/formats/test_ftl.py

```diff
@@ -0,0 +1,291 @@
+        save.
+        """
+        path = self.get_nonexistant_file_path()
+        translated_resource = self.get_nonexistant_file_resource(path)
+
+        translated_resource.translations[0].strings = {
+            None: 'New Translated String'
+        }
+        translated_resource.save(LocaleFactory.create())
+        assert_true(os.path.exists(path))
+
+    def test_parse_with_source_path(self):
+        contents = dedent("""text = Arise, awake and do not stop until the goal is reached.""")
+        source_path = create_named_tempfile(
+            contents,
+            prefix='strings',
+            suffix='.ftl',
+            directory=self.tempdir,
+        )
+        path = self.get_nonexistant_file_path()
+        obj = ftl.parse(path, source_path=source_path, locale=None)
+        assert_equal(obj.path, path)
+        assert_equal(obj.locale, None)
+        assert_equal(obj.source_resource.path, source_path)
+        assert_equal(obj.source_resource.locale, None)
+
+    def test_parse_with_no_source_path(self):
+        contents = dedent("""text = Arise, awake and do not stop until the goal is reached.""")
+        path = create_named_tempfile(
+            contents,
+            prefix='strings',
+            suffix='.ftl',
+            directory=self.tempdir,
+        )
+        obj = ftl.parse(path, source_path=None, locale=None)
+        assert_equal(obj.path, path)
+        assert_equal(obj.source_resource, None)
+        assert_equal(obj.locale, None)
+
+
+BASE_FTL_FILE = """
+# Sample comment
+SourceString = Translated String
+
+# First comment
+# Second comment
+MultipleComments = Translated Multiple Comments
+
+NoCommentsorSources = Translated No Comments or Sources
+"""
+
+
+class FTLTests(FormatTestsMixin, TestCase):
+    parse = staticmethod(ftl.parse)
+    supports_keys = False
+    supports_source = False
+    supports_source_string = False
+
+    def setUp(self):
+        super(FTLTests, self).setUp()
+        self.tempdir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        super(FTLTests, self).tearDown()
+        shutil.rmtree(self.tempdir)
+
+    def key(self, source_string):
+        """FTL keys can't contain spaces."""
+        return super(FTLTests, self).key(source_string).replace(' ', '')
+
+    def test_parse_basic(self):
+        input_string = BASE_FTL_FILE
+        translation_index = 0
+        path, resource = self.parse_string(input_string)
+
+        assert_attributes_equal(
+            resource.translations[translation_index],
+            comments=['Sample comment'],
+            key=self.key('Source String'),
+            strings={None: 'SourceString = Translated String\n'},
+            fuzzy=False,
+            order=translation_index,
+        )
+
+    def test_parse_multiple_comments(self):
+        input_string = BASE_FTL_FILE
+        translation_index = 1
+        path, resource = self.parse_string(input_string)
```

Comment:
Ouch, that hurts a bit. I'm not saying you should change it here, but this is a piece of complex code, that is very hard to understand without already knowing the code base. It depends on the structure of `FormatTestsMixin` a lot, and hides the function that is actually being tested (in this case, `ftl.parse`). To me this is a complexity that we should avoid in test (and in code in general) if possible, even if that means repeating code, in favor of making tests as straightforward as possible. Anyway, just ranting. :grin:

Reply:
Good point. We'll likely migrate most of the formats to `compare-locales`, which would be a good opportunity to make this code more explicit.